Bezier curves for automation tracks

codythecoder opened the pull request:
Based on [this](https://github.com/LMMS/lmms/issues/2381) issue, I think having bezier curves in LMMS would make automation tracks a lot more powerful, and much faster to do certain types of curves.

tresf commented:
Some notes... 

 1. The compile is failing
 2. Please squash your commit history
 3. Why are the mingw scripts changed in this?  Can you please explain or remove?

softrabbit commented:
Looks like the windows builds are failing, that's what you get when you remove the execute bit on build scripts...

codythecoder commented:
I've fixed the compile problems. I must've accidentally done something to the windows build scripts so I just copied the master branch's code over.
I'm not sure how to squash my commit history. I think I squashed it a bit, but now things aren't working and I'm not sure what to do. When I type `git rebase -i HEAD~4` it shows much more than 4 commits, and not the ones I want. If someone could help with this or point me in the right direction that would be great.

tresf commented:
@codythecoder first, backup the files you've changed.

Then rewind your branch.  One way to do this is a `git reset --hard abcdefg` where `abcdefg` is a [commit hash from the past](https://github.com/LMMS/lmms/commits?page=6) that you'd have in your fork.

Next, [fast-forward your branch to have the latest from upstream](https://github.com/LMMS/lmms/wiki/%5BGit%5D-Fast-forward-personal-branch-to-upstream).

Last, re-apply your changes manually.

That's not to say your existing commits can't be resurrected, but a bad rebase is hard to unwind, sometimes it's easier to just start from scratch. :+1:

codythecoder commented:
Can someone help me figure out what's going on with the build fails? It's giving me warnings about packages not being authenticated, but I'm not sure how to fix this or what I could have done to cause this. After this is solved, it should be ready to merge with master.

zonkmachine commented:
```make clean```?

tresf commented:
> Can someone help me figure out what's going on with the build fails? 

@codythecoder, I re-started the Travis-CI builds.  I don't think the errors are related to your commit specifically.  Let's see if they're in better shape in a few minutes here.

codythecoder commented:
Cool, so that works now. I just realized that I still have to change one thing, which I just have to test and commit, so that will be done in a few minutes.

codythecoder commented:
Ok, that's all ready :)

zonkmachine commented:
Build failed again. I restarted it but it wouldn't work anyway...

tresf commented:
Yeah, Travis is having some technical difficulties.  There's a [GPG key here](https://travis-ci.org/LMMS/lmms/jobs/104554441#L140) which is unreachable at the moment, causing some dependencies to fail, thus false-positively failing the tests.

This missing Canonical/Ubuntu GPG key is affecting all builds which are Ubuntu-derived (QT4, QT5, Win32, Win64).  Apple isn't affected since it uses Homebrew for all of its dependencies.  Ubuntu 12.04 ("Precise") isn't [EOL until April 2017](https://wiki.ubuntu.com/Releases) so the mirrors should be up for another 15 months.  Simply re-running the Travis build process again a bit later will likely resolve this. :+1:

tresf commented:
Travis-CI is passing. :+1:

codythecoder commented:
I'm all done with this, when can we get this merged?

grejppi commented:
@codythecoder Is it intentional that when you move a point its handles will reset?

codythecoder commented:
@grejppi Yes, when I wrote it, it was intentional. Getting the dragging to work was one of the first things I had to do, and due to how point dragging is normally done, I thought it would be impossible to get the control points to move along with the automation point, and I just forgot about it. 
After a bit of looking around, I think it might be possible to get the control points to move with the automation point in a way that feels more natural. I'll spend a couple of days to see if I can find a way to get this up.

codythecoder commented:
I've got it to drag the control points along with the automation points, I think that's everything, but if people could test it while I fix the conflicts, that would be great.

zonkmachine commented:
This pull request has merge conflicts and needs to be rebased against current master.

tresf commented:
> test it while I fix the conflicts, that would be great.

> > This pull request has merge conflicts and needs to be rebased against current master.

:laughing:

zonkmachine commented:
>>> test it while I fix the conflicts, that would be great.
>> 
>>         This pull request has merge conflicts and needs to be rebased against current master.
> 
> :laughing:

Still, it's well within my usual margin of error...


> but if people could test it

It crashes for me. Both rebased and in present form. I'll try and add some debug data later.

zonkmachine commented:
Crashes when I click in the autmation pattern.
```
ASSERT failure in QVector<T>::operator[]: "index out of range", file /usr/include/qt4/QtCore/qvector.h, line 359
Aborted (core dumped)
```

Backtrace
```
(gdb) bt
#0  0x00007fb04102ecc9 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007fb0410320d8 in __GI_abort () at abort.c:89
#2  0x00007fb043798c92 in qt_message_output(QtMsgType, char const*) () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#3  0x00007fb043798ff9 in ?? () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#4  0x00007fb043799804 in qFatal(char const*, ...) () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#5  0x00000000004e4fe5 in QVector<float>::operator[] (this=0x346f788, i=1) at /usr/include/qt4/QtCore/qvector.h:359
#6  0x00000000004de5bc in AutomationPattern::setDragValue (this=0x7fb037d7fea0, _time=..., _value=50,6000023, _quant_pos=true)
    at /home/zonkmachine/builds/lmms/lmms/src/core/AutomationPattern.cpp:344
#7  0x00000000005b3f80 in AutomationEditor::mousePressEvent (this=0x2c7ada0, mouseEvent=0x7fff826a08f0)
    at /home/zonkmachine/builds/lmms/lmms/src/gui/editors/AutomationEditor.cpp:561
#8  0x00007fb04407138b in QWidget::event(QEvent*) () from /usr/lib/x86_64-linux-gnu/libQtGui.so.4
#9  0x00007fb044021e2c in QApplicationPrivate::notify_helper(QObject*, QEvent*) () from /usr/lib/x86_64-linux-gnu/libQtGui.so.4
#10 0x00007fb0440285dd in QApplication::notify(QObject*, QEvent*) () from /usr/lib/x86_64-linux-gnu/libQtGui.so.4
#11 0x00007fb0438a54dd in QCoreApplication::notifyInternal(QObject*, QEvent*) () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#12 0x00007fb044027d93 in QApplicationPrivate::sendMouseEvent(QWidget*, QMouseEvent*, QWidget*, QWidget*, QWidget**, QPointer<QWidget>&, bool) ()
   from /usr/lib/x86_64-linux-gnu/libQtGui.so.4
#13 0x00007fb04409c9eb in ?? () from /usr/lib/x86_64-linux-gnu/libQtGui.so.4
#14 0x00007fb04409c289 in QApplication::x11ProcessEvent(_XEvent*) () from /usr/lib/x86_64-linux-gnu/libQtGui.so.4
#15 0x00007fb0440c3b32 in ?? () from /usr/lib/x86_64-linux-gnu/libQtGui.so.4
#16 0x00007fb0408e0e04 in g_main_context_dispatch () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#17 0x00007fb0408e1048 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#18 0x00007fb0408e10ec in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#19 0x00007fb0438d27a1 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#20 0x00007fb0440c3be6 in ?? () from /usr/lib/x86_64-linux-gnu/libQtGui.so.4
#21 0x00007fb0438a40af in QEventLoop::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#22 0x00007fb0438a43a5 in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#23 0x00007fb0438a9b79 in QCoreApplication::exec() () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#24 0x00000000004d8170 in main (argc=1, argv=0x7fff826a13b8) at /home/zonkmachine/builds/lmms/lmms/src/core/main.cpp:807
(gdb)
```

codythecoder commented:
@zonkmachine I'll have a look at that after I finish with the conflicts, since there's a chance it'll just fix itself automatically. What system are you running it on?

zonkmachine commented:
> @zonkmachine I'll have a look at that after I finish with the conflicts, since there's a chance it'll just fix itself automatically.

Sorry! I should have mentioned that this was done rebased. Maybe I should wait to test it more until you've updated so we're working on the same source? I'll try and get some data from 1f25d1c9ca4f97545cce9230b996eaa6c76aa922 instead.

> What system are you running it on?

Linux Mint 17.3

zonkmachine commented:
> I'll try and get some data from 1f25d1c instead.

Same result.

`AutomationPattern::setDragValue (this=0x7fcededf62a0, _time=..., _value=0,559000015, _quant_pos=true)`

The  `_time=...` looks like it should have a value.

codythecoder commented:
I think I know where the error is coming from, so I'll see if I can get it to crash for me when I have some time. My guess is that `setDragValue` is being called from `mousePressEvent` and for some reason is not being passed the time, so I'll check that out.

codythecoder commented:
Oops I pressed the wrong button :smile:

zonkmachine commented:
Here's an odd thing. If I start lmms and terminate it from the terminal so I preserve the recovery.mmp it will also crash, and with the same message, if I try to recover the project.

```
(gdb) bt
#0  0x00007f7db9f16cc9 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007f7db9f1a0d8 in __GI_abort () at abort.c:89
#2  0x00007f7dbc680c92 in qt_message_output(QtMsgType, char const*) () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#3  0x00007f7dbc680ff9 in ?? () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#4  0x00007f7dbc681804 in qFatal(char const*, ...) () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#5  0x00000000004e4fe5 in QVector<float>::operator[] (this=0x2a41308, i=0) at /usr/include/qt4/QtCore/qvector.h:359
#6  0x00000000004e1a8c in AutomationPattern::loadSettings (this=0x7f7db0c56e20, _this=...)
    at /home/zonkmachine/builds/lmms/lmms/src/core/AutomationPattern.cpp:775
#7  0x000000000054623e in SerializingObject::restoreState (this=0x7f7db0c56e40, element=...)
    at /home/zonkmachine/builds/lmms/lmms/src/core/SerializingObject.cpp:70
#8  0x0000000000517dbc in JournallingObject::restoreState (this=0x7f7db0c56e40, _this=...)
    at /home/zonkmachine/builds/lmms/lmms/src/core/JournallingObject.cpp:92
#9  0x00000000005552f2 in Track::loadSettings (this=0x7f7db0c520a0, element=...) at /home/zonkmachine/builds/lmms/lmms/src/core/Track.cpp:2114
#10 0x000000000054623e in SerializingObject::restoreState (this=0x7f7db0c520c0, element=...)
    at /home/zonkmachine/builds/lmms/lmms/src/core/SerializingObject.cpp:70
#11 0x0000000000517dbc in JournallingObject::restoreState (this=0x7f7db0c520c0, _this=...)
    at /home/zonkmachine/builds/lmms/lmms/src/core/JournallingObject.cpp:92
#12 0x0000000000548d44 in Song::loadProject (this=0x2392d80, fileName=...) at /home/zonkmachine/builds/lmms/lmms/src/core/Song.cpp:973
#13 0x0000000000548973 in Song::createNewProjectFromTemplate (this=0x2392d80, templ=...) at /home/zonkmachine/builds/lmms/lmms/src/core/Song.cpp:915
#14 0x00000000004d7e97 in main (argc=1, argv=0x7ffc76859168) at /home/zonkmachine/builds/lmms/lmms/src/core/main.cpp:753
(gdb)
```

codythecoder commented:
I don't know if this will fix anything, but try running this newer version.

zonkmachine commented:
> I don't know if this will fix anything...

Now it no longer crash when I edit the automation track but it does on reloading a project. The second crash above is still valid though. It seem like not all projects will crash on recover or reload from gui but so far I've failed to see a pattern in there.

codythecoder commented:
I can't get it to crash for me, and I'm running the same versions as you (although a different os), so I'm not quite sure where the problem lies. Can you save and load tracks with bezier curves correctly?

zonkmachine commented:
> Can you save and load tracks with bezier curves correctly?

No. I can edit and save but it will crash on a reload. I've suspected it could be on my side all the time but I've taken all precautions I know of. Cleared the build and target directories and tried and tried again. The first commit worked like a charm but the second, as well as the rebased one, don't.

codythecoder commented:
Are you able to open either of these files? And could you save and send me a file with bezier curves?
[projects.zip](https://github.com/LMMS/lmms/files/207687/projects.zip)

Also, is anyone else having issues with bezier curves?

zonkmachine commented:
> Are you able to open either of these files?

No.


[zonk.mmp.zip](https://github.com/LMMS/lmms/files/207694/zonk.mmp.zip)

codythecoder commented:
Your file worked fine for me. Could you change this line in src/core/AutomationPattern.cpp and tell me what the result is when you try to load a file? 
![temp](https://cloud.githubusercontent.com/assets/13900631/14372068/a2187ff6-fd80-11e5-9d92-17b3ff615747.png)

zonkmachine commented:
1. I can now open your project 'discrete.mmpz' while the other projects still crash.
2. When I activate bezier curves in the Automation Pattern it crashes with

```
(gdb) bt
#0  0x00007fa2d4dcecc9 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007fa2d4dd20d8 in __GI_abort () at abort.c:89
#2  0x00007fa2d7538c92 in qt_message_output(QtMsgType, char const*) () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#3  0x00007fa2d7538ff9 in ?? () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#4  0x00007fa2d7539804 in qFatal(char const*, ...) () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#5  0x00000000004e52f3 in QVector<float>::operator[] (this=0x7ffc46ee9800, i=0) at /usr/include/qt4/QtCore/qvector.h:355
#6  0x00000000004df160 in AutomationPattern::valueAt (this=0x7fa2cbb1a8a0, v=..., offset=0)
    at /home/zonkmachine/builds/lmms/lmms/src/core/AutomationPattern.cpp:505
#7  0x00000000004dfc08 in AutomationPattern::valuesAfter (this=0x7fa2cbb1a8a0, _time=...)
    at /home/zonkmachine/builds/lmms/lmms/src/core/AutomationPattern.cpp:552
#8  0x00000000005b6c29 in AutomationEditor::paintEvent (this=0x1e61dc0, pe=0x7ffc46ee9e80)
    at /home/zonkmachine/builds/lmms/lmms/src/gui/editors/AutomationEditor.cpp:1341
#9  0x00007fa2d7e11300 in QWidget::event(QEvent*) () from /usr/lib/x86_64-linux-gnu/libQtGui.so.4
```

codythecoder commented:
I've got a couple of tests I'd like you to run, could you copy these into `loadSettings` in `src/core/automationPatern.cpp`, and then tell me what gets printed? The first and last lines are exact duplicates of other lines, and should show you where the tests go.

1:

	else if( element.tagName() == "ctrlpnt" )
		{
			printf("Test 1\n"); // New
			m_controlPoints[0][0] = 10;
			m_controlPoints[0][1] = 0.5;
			printf("Test 2\n");
			m_controlPoints[element.attribute( "pos" ).toInt()][0]
				= element.attribute( "value1" ).toInt();
			m_controlPoints[element.attribute( "pos" ).toInt()][1]
				= element.attribute( "value2" ).toFloat();
			printf("Test 3\n");
		}
		else if( element.tagName() == "object" )

2:

		else if( element.tagName() == "ctrlpnt" ) // No "&& false"
		{
			printf("Test 1\n");
			m_controlPoints[element.attribute( "pos" ).toInt()].insert( 0, element.attribute( "value1" ).toInt() );
			m_controlPoints[element.attribute( "pos" ).toInt()].insert( 1, element.attribute( "value2" ).toFloat() );
			printf("Test 2\n");
			printf("Test 3: %f\n", m_controlPoints[element.attribute( "pos" ).toInt()][1]);
		}
		else if( element.tagName() == "object" )

3:

			else if( element.tagName() == "ctrlpnt" )
			{
			}
			else if( element.tagName() == "object" )
			{
				m_idsToResolve << element.attribute( "id" ).toInt();
			}
		}

		int len = _this.attribute( "len" ).toInt();
		if( len <= 0 )
		{
			len = length();
		}
		changeLength( len );
		generateTangents();

		printf("Test 1\n");
		clampControlPoints();
		printf("Test 2\n");
	}

I'm reasonably sure that the problem is with `m_controlPoints[x][y]`, but I don't know for sure, or how extensive the problem goes.

softrabbit commented:
My first impression: Undo needs some care and attention. If I undo adding a point, it works as expected in the other modes but not in Bezier mode, where the point remains, but is moved to 0 on the Y axis.

zonkmachine commented:
Opening your file bezier.mmpz

First test

```
ASSERT failure in QVector<T>::operator[]: "index out of range", file /usr/include/qt4/QtCore/qvector.h, line 359
Aborted (core dumped)

(gdb) bt
#0  0x00007f2f91056cc9 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007f2f9105a0d8 in __GI_abort () at abort.c:89
#2  0x00007f2f937c0c92 in qt_message_output(QtMsgType, char const*) () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#3  0x00007f2f937c0ff9 in ?? () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#4  0x00007f2f937c1804 in qFatal(char const*, ...) () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#5  0x00000000004e50d5 in QVector<float>::operator[] (this=0x1947348, i=0) at /usr/include/qt4/QtCore/qvector.h:359
#6  0x00000000004e1aa1 in AutomationPattern::loadSettings (this=0x7f2f87d96e20, _this=...)
    at /home/zonkmachine/builds/lmms/lmms/src/core/AutomationPattern.cpp:764
#7  0x000000000054632e in SerializingObject::restoreState (this=0x7f2f87d96e40, element=...)
    at /home/zonkmachine/builds/lmms/lmms/src/core/SerializingObject.cpp:70
#8  0x0000000000517eac in JournallingObject::restoreState (this=0x7f2f87d96e40, _this=...)
    at /home/zonkmachine/builds/lmms/lmms/src/core/JournallingObject.cpp:92
#9  0x00000000005553e2 in Track::loadSettings (this=0x7f2f87d920a0, element=...) at /home/zonkmachine/builds/lmms/lmms/src/core/Track.cpp:2114
#10 0x000000000054632e in SerializingObject::restoreState (this=0x7f2f87d920c0, element=...)
    at /home/zonkmachine/builds/lmms/lmms/src/core/SerializingObject.cpp:70
#11 0x0000000000517eac in JournallingObject::restoreState (this=0x7f2f87d920c0, _this=...)
    at /home/zonkmachine/builds/lmms/lmms/src/core/JournallingObject.cpp:92
#12 0x0000000000548e34 in Song::loadProject (this=0xe53880, fileName=...) at /home/zonkmachine/builds/lmms/lmms/src/core/Song.cpp:973
#13 0x0000000000596cb0 in MainWindow::openRecentlyOpenedProject (this=0x103f200, _action=0x1a99c90)
    at /home/zonkmachine/builds/lmms/lmms/src/gui/MainWindow.cpp:9
```

Second test no crash.
```

Test 1
Test 2
Test 3: 4,000000
Test 1
Test 2
Test 3: 4,000000
Test 1
Test 2
Test 3: 140,000000
Test 1
Test 2
Test 3: 100,000000
Test 1
Test 2
Test 3: 0,000000
Test 1
Test 2
Test 3: 0,722000
Test 1
Test 2
Test 3: 0,339000
Test 1
Test 2
Test 3: 0,487000
```

Third test
```

Test 1
Test 2
Test 1
Test 2
Test 1
Test 2
Test 1
Test 2
Test 1
Test 2
Test 1
ASSERT failure in QVector<T>::operator[]: "index out of range", file /usr/include/qt4/QtCore/qvector.h, line 359
Aborted (core dumped)


(gdb) bt
#0  0x00007f300f336cc9 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007f300f33a0d8 in __GI_abort () at abort.c:89
#2  0x00007f3011aa0c92 in qt_message_output(QtMsgType, char const*) () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#3  0x00007f3011aa0ff9 in ?? () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#4  0x00007f3011aa1804 in qFatal(char const*, ...) () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#5  0x00000000004e4e19 in QVector<float>::operator[] (this=0x292fc78, i=0) at /usr/include/qt4/QtCore/qvector.h:359
#6  0x00000000004e20d4 in AutomationPattern::clampControlPoints (this=0x7f30060828a0)
    at /home/zonkmachine/builds/lmms/lmms/src/core/AutomationPattern.cpp:853
#7  0x00000000004e1cc6 in AutomationPattern::loadSettings (this=0x7f30060828a0, _this=...)
    at /home/zonkmachine/builds/lmms/lmms/src/core/AutomationPattern.cpp:779
#8  0x0000000000546072 in SerializingObject::restoreState (this=0x7f30060828c0, element=...)
    at /home/zonkmachine/builds/lmms/lmms/src/core/SerializingObject.cpp:70
#9  0x0000000000517bf0 in JournallingObject::restoreState (this=0x7f30060828c0, _this=...)
    at /home/zonkmachine/builds/lmms/lmms/src/core/JournallingObject.cpp:92
#10 0x0000000000555126 in Track::loadSettings (this=0x7f30060826a0, element=...) at /home/zonkmachine/builds/lmms/lmms/src/core/Track.cpp:2114
#11 0x0000000000546072 in SerializingObject::restoreState (this=0x7f30060826c0, element=...)
    at /home/zonkmachine/builds/lmms/lmms/src/core/SerializingObject.cpp:70
#12 0x0000000000517bf0 in JournallingObject::restoreState (this=0x7f30060826c0, _this=...)
    at /home/zonkmachine/builds/lmms/lmms/src/core/JournallingObject.cpp:92
#13 0x000000000055456f in Track::create (element=..., tc=0x1e36ca0) at /home/zonkmachine/builds/lmms/lmms/src/core/Track.cpp:1979
#14 0x000000000055a22d in TrackContainer::loadSettings (this=0x1e36ca0, _this=...) at /home/zonkmachine/builds/lmms/lmms/src/core/TrackContainer.cpp:123
#15 0x0000000000546072 in SerializingObject::restoreState (this=0x1e36cc0, element=...)
    at /home/zonkmachine/builds/lmms/lmms/src/core/SerializingObject.cpp:70
#16 0x0000000000517bf0 in JournallingObject::restoreState (this=0x1e36cc0, _this=...)
    at /home/zonkmachine/builds/lmms/lmms/src/core/JournallingObject.cpp:92
#17 0x0000000000548d40 in Song::loadProject (this=0x1e36ca0, fileName=...) at /home/zonkmachine/builds/lmms/lmms/src/core/Song.cpp:998
#18 0x00000000005969f4 in MainWindow::openRecentlyOpenedProject (this=0x1e35370, _action=0x2be3170)
    at /home/zonkmachine/builds/lmms/lmms/src/gui/MainWindow.cpp:913
```

codythecoder commented:
@zonkmachine Great, Thanks :+1: 

That's pretty much what I expected, so I should be able to write a fix for this, which I'll commit in a couple of days.

@softrabbit Undo is failing for similar reasons to @zonkmachine's problems, and I have it working on my local version, which I'll commit in a couple of days.

codythecoder commented:
Ok, this should fix any crashing (although there's a chance it won't), as well as any problems with undo/redo/loading.

zonkmachine commented:
Nice. No crash so far and undo works too. Looks like the issues are fixed.

codythecoder commented:
Travis is now failing, although it doesn't look like my fault. Can someone look into this?

tresf commented:
@codythecoder looking into it now...

tresf commented:
It's green.  It had some validation issues with some packages.  Had to kick start it a few times. (You're right not your fault.)

codythecoder commented:
@tresf cool 👍 now I guess it's just wait to see if any more problems arise?

tresf commented:
@zonkmachine can merge if he's comfortable with it and has tested it. :+1:

zonkmachine commented:
New crash. I got it when looping a BBTrack over four bars and modulating the cutoff on a 3osc with a bezier curve.

```
(gdb) bt
#0  0x00007ffff410ecc9 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007ffff41120d8 in __GI_abort () at abort.c:89
#2  0x00007ffff684fc92 in qt_message_output(QtMsgType, char const*) () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#3  0x00007ffff684fff9 in ?? () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#4  0x00007ffff6850804 in qFatal(char const*, ...) () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#5  0x00000000004e5511 in QVector<float>::operator[] (this=0x7fffbe0c2b90, i=1) at /usr/include/qt4/QtCore/qvector.h:355
#6  0x00000000004df3cd in AutomationPattern::valueAt (this=0x7fffeaed5ae0, v=..., offset=117)
    at /home/zonkmachine/builds/lmms/lmms/src/core/AutomationPattern.cpp:502
#7  0x00000000004decfc in AutomationPattern::valueAt (this=0x7fffeaed5ae0, _time=...)
    at /home/zonkmachine/builds/lmms/lmms/src/core/AutomationPattern.cpp:440
#8  0x00000000004e21f2 in AutomationPattern::processMidiTime (this=0x7fffeaed5ae0, time=...)
    at /home/zonkmachine/builds/lmms/lmms/src/core/AutomationPattern.cpp:807
#9  0x0000000000605584 in AutomationTrack::play (this=0x7fffeaed1ca0, _start=..., _frames=41, _frame_base=215, _tco_num=-1)
    at /home/zonkmachine/builds/lmms/lmms/src/tracks/AutomationTrack.cpp:85
#10 0x0000000000547811 in Song::processNextBuffer (this=0xc43950) at /home/zonkmachine/builds/lmms/lmms/src/core/Song.cpp:394
#11 0x00000000005238ec in Mixer::renderNextBuffer (this=0xbc6460) at /home/zonkmachine/builds/lmms/lmms/src/core/Mixer.cpp:397
#12 0x00000000005252b3 in Mixer::fifoWriter::run (this=0xe2b3a0) at /home/zonkmachine/builds/lmms/lmms/src/core/Mixer.cpp:1014
#13 0x00007ffff685a32f in ?? () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#14 0x00007ffff7bc4182 in start_thread (arg=0x7fffbe0c3700) at pthread_create.c:312
#15 0x00007ffff41d247d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
(gdb) 
```

[bezier2fullbacktrace.txt.zip](https://github.com/LMMS/lmms/files/229747/bezier2fullbacktrace.txt.zip)

codythecoder commented:
@zonkmachine If you save and load does it still crash? If it does, could you send the file to me and I can check it out. Also, could you copy this code into its spot in `float AutomationPattern::valueAt( timeMap::const_iterator v, int offset ) const` and tell me the last couple of numbers it prints before it crashes?

		// to make up for their range being limited.
		if ( !m_controlPoints.contains( v.key() ) )
			printf("0\n");
		printf("1\n");
		int targetX1 = ( m_controlPoints[v.key()][0] - v.key() ) * 2;
		printf("2\n");
		int targetX2 = ( 3 * (v+1).key() - 2 * m_controlPoints[(v+1).key()][0] - v.key() );
		// The y values are the actual y values. Maybe this should be doubled, 
		// but it doesn't seem necessary to me.
		float targetY1 = m_controlPoints[v.key()][1];
		float targetY2 = 2*(v+1).value() - m_controlPoints[(v+1).key()][1];
		printf("3\n");

		// To find the y value on the curve at a certain x, we first have to find the t (between 0 and 1) that gives the x

zonkmachine commented:
> If you save and load does it still crash?

Yes

> If it does, could you send the file to me and I can check it out.

[beziernewcrash-01.mmp.zip](https://github.com/LMMS/lmms/files/239613/beziernewcrash-01.mmp.zip)

> Also, could you copy this code into its spot in float AutomationPattern::valueAt( timeMap::const_iterator v, int offset ) const and tell me the last couple of numbers it prints before it crashes?

It counts 1 2 3 and never  prints 0. It most often crash on 1 not as often on 2 and rarely on 3.

codythecoder commented:
Sorry for the silence, uni and whatnot got in the way.

If you go back to `float AutomationPattern::valueAt( timeMap::const_iterator v, int offset ) const` and paste the following code, what numbers (if any) does it crash on?

		// to make up for their range being limited.
		if (m_controlPoints[(v+1).key()].size() != 2)
			printf("size: %i\n", m_controlPoints[(v+1).key()].size());
		printf("1\n");
		int targetX1 = ( m_controlPoints[v.key()].at(0) - v.key() ) * 2;
		printf("2\n");
		int targetX2 = ( 3 * (v+1).key() - 2 * m_controlPoints[(v+1).key()].at(0) - v.key() );
		// The y values are the actual y values. Maybe this should be doubled,
		// but it doesn't seem necessary to me.
		float targetY1 = m_controlPoints[v.key()].at(1);
		float targetY2 = 2*(v+1).value() - m_controlPoints[(v+1).key()].at(1);
		printf("3\n");

		// To find the y value on the curve at a certain x, we first have to find the t (between 0 and 1) that gives the x

zonkmachine commented:
Hi, finally got the time to sit down and try this but unfortunately it works well... :unamused: 
I can't make it crash no longer. Not with your master or when I rebase it to latest master.

codythecoder commented:
It's only in programming that it can be unfortunate when something works :smile: I'll push these changes anyway, maybe someone else'll have the same problem later, and hopefully this'll fix anything that would've crashed.

zonkmachine commented:
```
zonkmachine@zonkmachine:~/builds/lmms/lmms/build$ ./lmms 
Notice: could not set realtime priority.
VST sync support disabled in your configuration
ASSERT failure in QVector<T>::at: "index out of range", file /usr/include/qt4/QtCore/qvector.h, line 351
Aborted (core dumped)
```
:+1:  Crashed when manipulating one of the control points. Just looping the track worked well. 

[full backtrace](https://gist.github.com/zonkmachine/f90214df2ce9dc337e2f299413d8a051#file-backtrace-full-txt)
```
zonkmachine@zonkmachine:~/builds/lmms/lmms/build$ gdb ./lmms core 
...
[New LWP 6148]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
bCore was generated by `./lmms'.
Program terminated with signal SIGABRT, Aborted.
#0  0x00007fae71e96c37 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
56	../nptl/sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  0x00007fae71e96c37 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007fae71e9a028 in __GI_abort () at abort.c:89
#2  0x00007fae74600c92 in qt_message_output (msgType=msgType@entry=QtFatalMsg, 
    buf=0x7fae3001fa98 "ASSERT failure in QVector<T>::at: \"index out of range\", file /usr/include/qt4/QtCore/qvector.h, line 351")
    at global/qglobal.cpp:2383
#3  0x00007fae74600ff9 in qt_message(QtMsgType, const char *, typedef __va_list_tag __va_list_tag *) (msgType=msgType@entry=QtFatalMsg, 
    msg=0x7fae7476ea48 "ASSERT failure in %s: \"%s\", file %s, line %d", ap=ap@entry=0x7fae3b7949e8) at global/qglobal.cpp:2429
#4  0x00007fae74601804 in qFatal (msg=<optimized out>) at global/qglobal.cpp:2612
#5  0x00000000004e9e49 in QVector<float>::at (this=0x7fae3b794b80, i=1) at /usr/include/qt4/QtCore/qvector.h:351
#6  0x00000000004e3cff in AutomationPattern::valueAt (this=0x7fae68be6fa0, v=..., offset=24)
    at /home/zonkmachine/builds/lmms/lmms/src/core/AutomationPattern.cpp:504
#7  0x00000000004e362e in AutomationPattern::valueAt (this=0x7fae68be6fa0, _time=...)
    at /home/zonkmachine/builds/lmms/lmms/src/core/AutomationPattern.cpp:442
#8  0x00000000004e6b2a in AutomationPattern::processMidiTime (this=0x7fae68be6fa0, time=...)
    at /home/zonkmachine/builds/lmms/lmms/src/core/AutomationPattern.cpp:810
#9  0x000000000060cac4 in AutomationTrack::play (this=0x7fae68be6da0, _start=..., _frames=252, _frame_base=4, _tco_num=-1)
    at /home/zonkmachine/builds/lmms/lmms/src/tracks/AutomationTrack.cpp:85
#10 0x000000000054c28b in Song::processNextBuffer (this=0x1ef1fd0) at /home/zonkmachine/builds/lmms/lmms/src/core/Song.cpp:395
#11 0x000000000052864c in Mixer::renderNextBuffer (this=0x1e76660) at /home/zonkmachine/builds/lmms/lmms/src/core/Mixer.cpp:415
#12 0x000000000052a0e5 in Mixer::fifoWriter::run (this=0x1eef580) at /home/zonkmachine/builds/lmms/lmms/src/core/Mixer.cpp:1081
#13 0x00007fae7460b32f in QThreadPrivate::start (arg=0x1eef580) at thread/qthread_unix.cpp:349
#14 0x00007fae75980184 in start_thread (arg=0x7fae3b795700) at pthread_create.c:312
#15 0x00007fae71f5a37d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

zonkmachine commented:
@codythecoder This PR has conflicts (Probably Since https://github.com/LMMS/lmms/commit/41b930eb66ac05340ea5a8f120ee7090bb18886e).

Since it looks like it was mostly I who had issues with this PR, maybe it should be tested by someone else too? I'm off course happy to test anything you throw at me but I'd prefer it as a patch (`git diff > codeToTest.patch`) rather than instructions on what changes to apply.

jrcxyz commented:
What's the news on this? This would be such an amazing feature!! Glad to see it's still in the works anywho, and thanks for the effort toward something that would be so helpful :D :dango:

zonkmachine commented:
> What's the news on this?

That you just volonteered to test this PR.  :smiley:

codythecoder commented:
News of this has been that I've been busy (and not to mention lazy)  but it's probably time I finish it once and for all 😁

The only known problems remaining are: moving the control points while playing through the track will occasionally cause it to crash, and now of course it has merge conflicts.

I'll see what I can get done by the end of the week, but if you can find any other errors that's also good.

PhysSong commented:
@codythecoder Are you still here? If you are, could you continue your work?
>moving the control points while playing through the track will occasionally cause it to crash

Confirmed. If the OP @codythecoder is not here, maybe I can investigate it and open a new PR.
Anyway, this is great. Thank you for this work.

codythecoder commented:
I'll see what I can do. I'll get you more details once I get the workflow setup on my new computer.

PhysSong commented:
**Pro Tip**: Try to use `QPair` for control points, instead of `QVector` of size 2.  
FYI, I've succeed to resolve conflicts and other functions works well.

PhysSong commented:
@codythecoder Are you willing to continue working on this pull request?

codythecoder commented:
I'll have a good look tomorrow, but I imagine so

PhysSong commented:
@codythecoder Then may I continue your work?

PhysSong commented:
Continued in #4674. @codythecoder Thank you for the code!
